bip-mmhf.mediawiki

```diff
@@ -42,37 +42,72 @@ FIXME: SigOps within new blocks are to be counted only in witness scripts and th
 === Merkle tree algorithm ===
 
 For each object, compute a hash which represents it in the merkle tree (eg, txid or hash of the full transaction).
-Arrange these hashes into a vector.
 
-So long as the vector is more than one element, compress each set of two elements by replacing them with a SHA256d hash of:
-
-* 216 bits: leading 216 bits of the first element
-* 216 bits: leading 216 bits of the second element
-* 1 bit: element type
-* 1 bit: plurality (usually 1; see below)
-* 1 bit: merkle root (always 0 here)
-* 5 bits: always zero
-
-When the vector has an odd number of elements, the final element should be compressed with a SHA256d hash of:
-
-* 216 bits: leading 216 bits of the only remaining element
-* 216 bits: all zero
-* 1 bit: element type
-* 1 bit: plurality (0 in this case)
-* 1 bit: merkle root (always 0 here)
-* 5 bits: always zero
-
-For the first iteration of the compression loop (that is, when the elements being compressed are the object hashes themselves), the element type should be 1; for subsequent iterations, it must always be 0.
-
-Finally, to create the final merkle tree root, one last SHA256d hash is performed of:
-
-* 256 bits: complete remaining element
-* 64 bits: total number of elements in vector, encoded as big endian
-* 112 bits: all zero
-* 1 bit: element type (iff the vector was 1 item long, this will be 1)
-* 1 bit: plurality (always 0 here)
-* 1 bit: merkle root (always 1 here)
-* 5 bits: always zero
+The transactions are hashed using a Merkle sum-tree.  For the sum-tree, each hash consists of 5 fields.
+
+* 32 bytes: hash : The SHA256d of the hashed data
+* 8 bytes: fees : Total fees
+* 4 bytes: sigops : Total sigops
+* 8 bytes: size : Total size
+* 8 bytes: cost : Total cost
+
+For the flags byte, there are 3 bits defined.
+    
+    FLAG_LEAF = 0x80
+    FLAG_PLURAL = 0x40
+    FLAG_MERKLE_ROOT = 0x20
+    
+The remaining bits are reserved and should be set to zero.
+
+The FLAG_LEAF bit should be set for the lowest level of the tree only.
+
+The FLAG_PLURAL bit should be set when compressing two hashes into a single hash.
+
+The FLAG_MERKLE_ROOT bit should be set for the final hashing step to generate the merkle root.
+
+Arrange the hashes into a vector.
+
+For each pair of hashes, (left, right), in the vector, compress them by replacing them with:
+    
+    flags = FLAG_PLURAL (or FLAG_PLURAL | FLAG_LEAF)
```

Comment:
The transactions digest id defined as follows.  

tx.hash = SHA256d(tx serialization), i.e. txid
tx.fee =sum(inputs) - sum(outputs)
tx.sigops = number of sigops
tx.size = length(tx serialization)
tx.cost = size(witness) + 4 * size(tx serialization)

You create an array of those as your leaves of the tree.  For the first pass, FLAG_LEAF is set to 1.  For all the remaining ones it is set to zero.

Thinking about it, the cost sum-tree may not be appropriate for the main Merkle tree.  Since it includes the size of the witness, you can't verify that by looking at the transaction.

The witness tree should have the sum of the witness sizes.